### PR TITLE
Infra mvp

### PR DESCRIPTION
**Pull request by JaafarRammal**
# Summary

MVP ready for main

## Additional Information

- Tower
- Agents
- Simen
- Message passing
- Tower interaction

## Test Plan

See #16 for future as well as verified mutex logic for async accesses

### CODE REVIEW

File: pkg/agents/team1/agent1/agent.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"log"
+
+	"github.com/SOMAS2021/SOMAS2021/pkg/agents"
+	"github.com/SOMAS2021/SOMAS2021/pkg/infra/messages"
+	"github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
+)
+
+type CustomAgent1 struct {
+	*agents.Base
+	myNumber int
```
Comment (tomeaton17) [inline pkg/agents/team1/agent1/agent.go:13]: What is the point of myNumber? Better naming would clarify. Seems like an ID but not obvious.

File: pkg/agents/team2/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+<!-- Team 1 packages -->
```
Comment (tomeaton17) [inline pkg/agents/team2/README.md:1]: Very minor, but incorrect number of teams.

File: pkg/agents/baseagent.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+func (a *Base) HP() int {
+	return a.tower.HP(a.id)
+}
+
+func (a *Base) Floor() int {
+	return a.tower.Floor(a.id)
+}
+
+func (a *Base) ID() string {
```
Comment (hussain2603) [inline pkg/agents/baseagent.go:45]: very minor, but maybe should name the HP() function 'get_HP()' just to make it a little clearer. Same for Floor() and ID().

File: pkg/infra/tower/tower-agent.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	_, found := tower.agents[id]
+	return found
+}
+
+func (tower *Tower) setFloor(id string, newFloor int) {
+	tower.mx.Lock()
+	agent := tower.agents[id]
+	agent.floor = newFloor
+	tower.agents[id] = agent
+	tower.mx.Unlock()
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-agent.go:39]:
```suggestion
	tower.mx.Lock()
	defer tower.mx.Unlock()
	agent := tower.agents[id]
	agent.floor = newFloor
	tower.agents[id] = agent
```

File: pkg/infra/tower/tower-agent.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func (tower *Tower) setHP(id string, newHP int) {
+	tower.mx.Lock()
+	agent := tower.agents[id]
+	agent.hp = newHP
+	tower.agents[id] = agent
+	tower.mx.Unlock()
+}
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-agent.go:48]:
```suggestion
	tower.mx.Lock()
	defer tower.mx.Unlock()
	agent := tower.agents[id]
	agent.hp = newHP
	tower.agents[id] = agent
}
```

File: pkg/infra/tower/tower-functions.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+}
+
+func (t *Tower) updateHP(id string, foodTaken float64) {
+	t.mx.Lock()
+	newHP := math.Min(100, float64(t.agents[id].hp)+foodTaken)
+	t.mx.Unlock()
+	t.setHP(id, int(newHP))
+}
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:70]:
```suggestion
func (t *Tower) updateHP(id string, foodTaken float) {
	t.mx.Lock()
	newHP := math.Min(100, float(t.agents[id].hp)+foodTaken)
	t.mx.Unlock()
	t.setHP(id, int(newHP))
}
```

File: pkg/infra/tower/tower-functions.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+func (t *Tower) FoodRequest(id string, foodRequested float64) float64 {
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:72]:
```suggestion
func (t *Tower) FoodRequest(id string, foodRequested float) float {
```

File: pkg/infra/tower/tower-functions.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}
+	for _, agent := range tower.agents {
+		//find reciever and pass them msg
+		if agent.floor == senderFloor+direction {
+			agent.inbox.PushBack(msg) //<- msg //for some reason channels were causing hanging
+		}
+	}
+	tower.mx.Unlock()
+}
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:107]:
```suggestion
func (tower *Tower) SendMessage(direction int, sender abm.Agent, msg messages.Message) {
	tower.mx.Lock()
	defer tower.mx.Unlock()
	senderFloor := 0
	for id, agent := range tower.agents {
		//find sender BaseAgentCore
		if id == (sender).ID() {
			senderFloor = agent.floor
		}
	}
	for _, agent := range tower.agents {
		//find receiver and pass them msg
		if agent.floor == senderFloor + direction {
			agent.inbox.PushBack(msg) //<- msg //for some reason channels were causing hanging
		}
	}
}
```

File: pkg/infra/tower/tower-functions.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+			if agent.inbox.Len() > 0 {
+				front := agent.inbox.Front()
+				msg := (front.Value.(messages.Message)) //<-agent.inbox:
+				(agent.inbox).Remove(front)
+				return msg
+			}
+		}
+	}
+	return nil
+}
```
Comment (jzzheng22) [inline pkg/infra/tower/tower-functions.go:124]:
```suggestion
func (tower *Tower) ReceiveMessage(receiver abm.Agent) messages.Message {
	tower.mx.Lock()
	defer tower.mx.Unlock()
	for id, agent := range tower.agents {
		//find sender BaseAgentCore
		if id == (receiver).ID() {
			if agent.inbox.Len() > 0 {
				front := agent.inbox.Front()
				msg := (front.Value.(messages.Message)) //<-agent.inbox:
				(agent.inbox).Remove(front)
				return msg
			}
		}
	}
	return nil
}
```

File: pkg/infra/tower/tower.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
+package tower
+
+import (
+	"log"
+	"sync"
+)
+
+type Tower struct {
+	currPlatFood    float64
+	maxPlatFood     float64
```
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:10]:
```suggestion
	currPlatFood    float
	maxPlatFood     float
```

File: pkg/infra/tower/tower.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	agentCount      int
+	agents          map[string]BaseAgentCore // TODO: change this to a map of pointers to structs
+	agentsPerFloor  int
+	ticksPerDay     int
+	missingAgents   map[int][]int // key: floor, value: types of missing agents
+	reshufflePeriod int
+	tickCounter     int
+}
+
+func New(currPlatFood float64, currPlatFloor int, agentCount,
```
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:22]:
```suggestion
func New(currPlatFood float, currPlatFloor int, agentCount,
```

File: pkg/infra/tower/tower.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	log.Printf("A log from the tower! Tick no: %d", t.tickCounter)
+	log.Printf("The food left on the platform = %f", t.currPlatFood)
+
+	//useful parameters
+	day := 24 * 60
+	numOfFloors := t.agentCount / t.agentsPerFloor
+	platformMovePeriod := day / numOfFloors // can add min/max
+
+	// Shuffle the agents
+	if t.tickCounter%t.reshufflePeriod == 0 {
```
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:53]:
```suggestion
	if t.tickCounter % t.reshufflePeriod == 0 {
```

File: pkg/infra/tower/tower.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+		t.mx.Unlock()
+		t.reshuffle(numOfFloors)
+		t.mx.Lock()
+	}
+	// Move the platform
+	if (t.tickCounter)%(platformMovePeriod) == 0 {
```
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:59]:
```suggestion
	if  t.tickCounter % platformMovePeriod == 0 {
```

File: pkg/infra/tower/tower.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+		t.currPlatFloor++
+	}
+	// Decrease agent HP and reset tower at end of day
+	if (t.tickCounter)%(day) == 0 {
```
Comment (jzzheng22) [inline pkg/infra/tower/tower.go:63]:
```suggestion
	if t.tickCounter % day == 0 {
```

File: pkg/utils/abm/abm.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+	currIteration int
+	limit         int
+
+	world      World
+	reportFunc func(*ABM)
+}
+
+// New creates new ABM simulation engine with default
+// parameters.
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:21]:
```suggestion
// New creates new ABM simulation engine with default parameters.
```

File: pkg/utils/abm/abm.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+// Iteration returns current iteration (age, generation).
+func (a *ABM) Iteration() int {
+	a.mx.Lock()
+	defer a.mx.Unlock()
+	return a.currIteration
+}
+
+func (a *ABM) SimulationIterate(i int) {
+
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:92]:
```suggestion
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: pkg/agents/baseagent.go (hunk 3 above)
Comment (jzzheng22) [reply pkg/agents/baseagent.go:45]: Please see this for why it is just `HP()`: https://go.dev/doc/effective_go#Getters

File: pkg/infra/simulation/simulation.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+		}
+	}
+	a.LimitIterations(sE.Iterations)
+	sE.simulationLoop(a, t)
+}
+
+func (sE *SimEnv) simulationLoop(a *abm.ABM, t *tower.Tower) {
+	// TODO: abstract with an extra function
+	for i := 1; i <= sE.Iterations; i++ {
+		missingAgentsMap := t.UpdateMissingAgents()
```
Comment (hussain2603) [inline pkg/infra/simulation/simulation.go:70]: I feel like this bit should probably be done elsewhere in a separate function called 'replace_dead_Agents' or smth

Comment (jzzheng22) [reply pkg/infra/simulation/simulation.go:70]: See #20. Will be resolved post-MVP release.

File: pkg/infra/tower/tower.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
+package tower
+
+import (
+	"log"
+	"sync"
+)
+
+type Tower struct {
+	currPlatFood    float64
```
Comment (hussain2603) [inline pkg/infra/tower/tower.go:9]: I don't think we would 64 bits to specify the amount of Food ever, would we?

File: pkg/infra/messages/foodRecentlyEatenMessage.go [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package messages
+
+//The last time an agent eats this is how much food they ate
+type FoodRecentlyEatenMessage struct {
+	*baseMessage
+	FoodAmt uint
```
[the remark this answers is not in the document]
Comment (tomeaton17) [reply pkg/infra/messages/foodRecentlyEatenMessage.go:6]: What is the downside to using uint? I think that Uint improves readability and makes it clear that FoodAmt >= 0.

[the remark this answers is not in the document]
Comment (jzzheng22) [reply pkg/infra/messages/foodRecentlyEatenMessage.go:6]: See https://github.com/golang/go/issues/27460, https://stackoverflow.com/questions/30395205/why-are-unsigned-integers-error-prone for examples

[the remark this answers is not in the document]
Comment (tomeaton17) [reply pkg/infra/messages/foodRecentlyEatenMessage.go:6]: Most of the examples here result from bad programming practices. However, if we would like to provide extra security against this then we can change to int.

File: pkg/agents/team1/agent1/agent.go (hunk 1 above)
Comment (JaafarRammal) [reply pkg/agents/team1/agent1/agent.go:13]: Example of a custom parameter you can add in

**Comment by JaafarRammal** [comment]
#23 on the way instead with changes from here and some refactors. Stay tuned

**Comment by JaafarRammal** [comment]
#23 on the way instead with changes from here and some refactors. Stay tuned

**Pull request closed without merging**